server/build_event_protocol/event_index/event_index.go

```diff
@@ -0,0 +1,127 @@
+package event_index
+
+import (
+	"sort"
+
+	cmnpb "github.com/buildbuddy-io/buildbuddy/proto/api/v1/common"
+	bespb "github.com/buildbuddy-io/buildbuddy/proto/build_event_stream"
+	inpb "github.com/buildbuddy-io/buildbuddy/proto/invocation"
+	trpb "github.com/buildbuddy-io/buildbuddy/proto/target"
+	api_common "github.com/buildbuddy-io/buildbuddy/server/api/common"
+)
+
+// Index holds a few data structures to make it easier to aggregate data from
+// raw BES events and organize them into pages.
+type Index struct {
+	AllTargetLabels            []string
+	BuildTargetByLabel         map[string]*trpb.Target
+	TestTargetByLabel          map[string]*trpb.Target
+	TargetCompleteEventByLabel map[string]*bespb.TargetComplete
+	TestResultEventsByLabel    map[string][]*bespb.BuildEvent
+	TargetsByStatus            map[cmnpb.Status][]*trpb.Target
+	NamedSetOfFilesByID        map[string]*bespb.NamedSetOfFiles
+	// Events which aren't indexed by target and are instead returned in the
+	// top-level invocation proto.
+	TopLevelEvents []*inpb.InvocationEvent
+
+	rootCauseLabels map[string]bool
+}
+
+func Build(in *inpb.Invocation) *Index {
```

Comment:
I don't think we want this API to take a full invocation because that implies we're buffering a full invocation in memory, which we want to stop doing.

Instead can we parse events as we read them from blobstore (or from the invocation stream, if you're accumulating these eventually)?

Like, the caller of event_index can use:
```
func streamRawInvocationEvents(env environment.Env, ctx context.Context, streamID string, callback invocationEventCB) error
```

Reply:
I plan to do this in a follow-up PR, since it will involve tinkering with the LookupInvocation API in a  way that I want to think through carefully.